Custom auth/alban/react next sample app signin

albanx opened the pull request:
1. Completed Sign In sample with NextJs
2. [Bug] Created Sign Up but not working due to missing exposure of submitCode in the result.state
3. Removed correlationId from Fetch client because it is present already in the header
4. Removed HTTPS checks in SDK, it is not needed to be handled by JS and blocks local development

Commit:
Skeleton

Commit:
Commit

Commit:
Commit

Commit:
Commit

shenj committed:
Refactor 1

shenj committed:
Update for design doc

shenj committed:
Commit

shenj committed:
rename design doc

shenj committed:
Design update

shenj committed:
Commit

shenj committed:
Remove the design doc

shenj committed:
ignore build output

shenj committed:
Updates for feedbacks

shenj committed:
Commit

shenj committed:
Commit

shenj committed:
Restructure folders

shenj committed:
rename package

shenj committed:
Rename

shenj committed:
feedbacks

nguyencuong2596 committed:
- Setup msal-custom-auth sample for e2e
- Add msal-custom-auth to pipeline

nguyencuong2596 committed:
Merge branch 'jiashen/sketeton' of https://github.com/AzureAD/microsoft-authentication-library-for-js into jiashen/sketeton

shenj committed:
Commit

nguyencuong2596 committed:
Merge branch 'dev' into jiashen/sketeton

nguyencuong2596 committed:
Change build command

nguyencuong2596 committed:
Update packages

nguyencuong2596 committed:
Alter build:all

nguyencuong2596 committed:
Change files

nguyencuong2596 committed:
Fix lint and alter build:all

nguyencuong2596 committed:
Change package

nguyencuong2596 committed:
Modify build:all

nguyencuong2596 committed:
Update package lock

nguyencuong2596 committed:
Update package

nguyencuong2596 committed:
Update package

nguyencuong2596 committed:
Update readme

nguyencuong2596 committed:
Update package.json

nguyencuong2596 committed:
Change files

nguyencuong2596 committed:
Update sample-sample

nguyencuong2596 committed:
Add puppeteer and e2e-test-utils

nguyencuong2596 committed:
Format code

nguyencuong2596 committed:
Use shx rm instead of rimraf

nguyencuong2596 committed:
Update api document

shenj committed:
Common changes for signin

axhaferllari committed:
Adding Sign up Api Client and Sign Up types

axhaferllari committed:
Remove unused file

shenj committed:
Sign-in completed updated

shenj committed:
Fix the style error

shenj committed:
Merge pull request #7508 from AzureAD/custom-auth/jiashen/signin

Sign-in completed updated

shenj committed:
Create the interfaces for signup in network client (#7521)

shenj committed:
Custom auth/jiashen/sign up (#7542)

shenj committed:
Add unit tests for sign-up

albanx committed:
Add Api Clients for all endpoints (#7553)

Adding network clients

---------

Co-authored-by: Alban Xhaferllari <axhaferllari@microsoft.com>

shenj committed:
Fix the network client issue.

shenj committed:
Fix the format issue

shenj committed:
Reset password feature implementation

shenj committed:
Add unit tests for password reset

shenj committed:
Add client_info property when calling signin token endpoint

shenj committed:
Support cache entities generation and caching

axhaferllari committed:
[CustomAuth] Create React JS Sample skeleton

axhaferllari committed:
remove nextjs images

axhaferllari committed:
add npm install

axhaferllari committed:
Update instructions and docs

axhaferllari committed:
Fix override interfeer

shenj committed:
GetCurrentAccount/SignOut implementation and a few improvements (#7586)

yongdiw committed:
Support get access token

yongdiw committed:
Remove msal-common references and create same function for generating expireOn for AuthorityMetadataEntity

yongdiw committed:
Remove package.json changes

yongdiw committed:
Fix comments

yongdiw committed:
Fix comments - moving functions, add default scopes, logs

yongdiw committed:
Fix comments - remove isRefreshTokenValid, remove un-needed error handling

yongdiw committed:
Fix comments - move createCommonSilentFlowRequest, fix its unit tests

shenj committed:
Merge branch 'dev' into custom-auth/main

shenj committed:
Fix the merge issue

yongdiw committed:
Fix comments - remove unused idTokenHints constants and invalidScopes constants

yongdiw committed:
Fix comments - update AuthorityMetadataEntity expireAt, update for getAccessToken getCurrentAccount

yongdiw committed:
Fix comments - give max safe integer

yongdiw committed:
Remove unused constants and references

yongdiw committed:
Merge pull request #7595 from AzureAD/custom-auth/get-token

Support get access token and token renewal

axhaferllari committed:
Changing to yarn to avoid conflicts

axhaferllari committed:
update readme

axhaferllari committed:
Merge branch 'custom-auth/main' into custom-auth/alban/react-nextjs-sample-skeleton

# Conflicts:
#	package-lock.json

axhaferllari committed:
add base SignIn form with config _not working_

axhaferllari committed:
Fix: import Constant from msal-common and update version of msal-browser to fix build

axhaferllari committed:
Merge branch 'custom-auth/main' into custom-auth/alban/react-nextjs-sample-skeleton

axhaferllari committed:
Add forms

shenj committed:
Fix the errors introduced by msal-browser breaking changes (#7605)

yongdiw committed:
Update Error handling and log pii info

yongdiw committed:
comments - remove public key word

yongdiw committed:
Comments - add correlation id for all logs

yongdiw committed:
Comments - add subError property

yongdiw committed:
Merge pull request #7608 from AzureAD/yongdi/error-handling-pii

Update Error handling and log pii info

shenj committed:
Some improvements (#7607)

axhaferllari committed:
refactor

axhaferllari committed:
update pipeline config

axhaferllari committed:
add a sample test

albanx committed:
Merge pull request #7583 from AzureAD/custom-auth/alban/react-nextjs-sample-skeleton

[CustomAuth] Create React JS Sample skeleton
will address Jian comment in the next merge

yongdiw committed:
Update logging level to use verbose and update index.ts (#7617)

This PR is for:
1. Updating all the log level to use verbose for better debug purposes.
2. Update index.ts to expose sdk components

axhaferllari committed:
Complete Sign In
Coded Sign Up but not working

axhaferllari committed:
remove custom tenant config

axhaferllari committed:
Add Angular sample

axhaferllari committed:
Merge branch 'custom-auth/main' into custom-auth/alban/react-next-sample-app-signin

# Conflicts:
#	lib/msal-custom-auth/src/core/network_client/http_client/FetchHttpClient.ts

shenj commented on lib/msal-custom-auth/.editorconfig:
```diff
@@ -6,7 +6,7 @@ end_of_line = lf
 trim_trailing_whitespace = true
 insert_final_newline = true
-max_line_length = 120
+max_line_length = 140
```
I will suggest using 120 since some people like open two windows side by side in a big screen.

shenj commented on lib/msal-custom-auth/src/CustomAuthPublicClientApplication.ts:
```diff
@@ -125,12 +125,5 @@ export class CustomAuthPublicClientApplication
-                InvalidAuthApiProxyDomain,
-                `The authApiProxyDomain URL '${config.customAuth.authApiProxyUrl}' is not a valid secure URL.`,
-            );
-        }
```
Why this is removed?

albanx replied:
blocking local dev. maybe we can allow localhost URLs for developers. I am of the idea though that there is no value in checking if a URL is https on front end code

shenj commented on lib/msal-custom-auth/src/core/network_client/custom_auth_api/BaseApiClient.ts:
```diff
@@ -18,7 +18,7 @@ export abstract class BaseApiClient {
         private httpClient: IHttpClient,
     ) {
-        this.baseRequestUrl = UrlUtils.parseSecureUrl(!baseUrl.endsWith("/") ? `${baseUrl}/` : baseUrl);
+        this.baseRequestUrl = UrlUtils.parseUrl(!baseUrl.endsWith("/") ? `${baseUrl}/` : baseUrl);
```
This is suggestion from security team. We need to ensure the api url is using https.

albanx replied:
We need support for local dev though. Also this check should be done in API not in front end. Doing it on JS is a nice to have but not a security measure.

shenj replied:
So, please keep it if it is nice to have. If you have any problem for the local dev, let us consider a solution.

albanx replied:
No it blocks the dev development and it does not bring any security feature on front end. this should be removed.

shenj commented on lib/msal-custom-auth/src/core/network_client/http_client/FetchHttpClient.ts:
```diff
@@ -13,14 +13,15 @@ import { FailedSendRequest, HttpError, NoNetworkConnectivity } from "../../error
-    async sendAsync(url: string | URL, options: RequestInit, correlationId: string): Promise<Response> {
+    async sendAsync(url: string | URL, options: RequestInit): Promise<Response> {
+        const headers = options.headers as Record<string, string>;
+        const correlationId = headers?.["client-request-id"] || undefined;
```
Please use the constant `AADServerParamKeys.CLIENT_REQUEST_ID` as header key.

albanx replied:
Good call.

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-in/page.tsx:
```diff
@@ -1,10 +1,117 @@
+                password,
+            });
+
+            if (result.error) {
```
IMO, it is better to use following check in the if (xxx).
`if (result.state?.type === SignInState.Failed)`

albanx replied:
If I check state as above this causes a type check error (for the reasons widely discussed in the chat today). This happens because the result.error is ambiguous due to the polymorphic type of `result`:


![image](https://github.com/user-attachments/assets/2256ef6e-8262-478c-8597-0effe16b7c8b)

![image](https://github.com/user-attachments/assets/e8304f13-ecbd-4154-9ccc-6ca015949117)

shenj replied:
I didn't see any problem.

<img width="514" alt="image" src="https://github.com/user-attachments/assets/fcb274bf-85f1-4aaf-b0c0-7f2b8171581a" />

albanx replied:
Check SignUp return result not signIn

shenj replied:
Here is the signup:
![image](https://github.com/user-attachments/assets/c54a97bd-4ad4-44a9-8716-84d4bb9b812f)

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-in/page.tsx:
```diff
@@ -1,10 +1,117 @@
+        setError("");
+        setLoading(true);
+
+        try {
```
There is no need the type catch here since the SDK is guaranteed no error thrown

albanx replied:
there is a `await` promise resolve there, meaning that the try catch is a must of the promise rejects

shenj replied:
Afaik, the promise rejection only occurs if an error is thrown and not caught inside the method. But as I said, any API in this SDK is guaranteed no error thrown.

albanx replied:
A JavaScript promise has 2 statuses: resolve, reject. 
The compiler or javascript cannot know that the SDK internally handles all errors of that away reject.
Without the try-catch, any error of the promise would cause the function to immediately throw an exception and potentially crash.

albanx replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#handling_rejected_promises

shenj replied:
I didn't get it. If we guarantee the API will never reject the promise or throw an error, is the try catch still necessary?

albanx replied:
You can guarantee it, but JavaScript engine cannot know it. No developer can know it, unless the read the source of the SDK itself.
If I read `await something` -> that means it can throw an exception.

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-in/page.tsx:
```diff
@@ -1,10 +1,117 @@
+        setError("");
+        setLoading(true);
+
+        try {
```
Same, no need try catch here.

albanx replied:
same as above, JS await handling promises

shenj replied:
Please check my reply above.

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-up/page.tsx:
```diff
@@ -1,10 +1,117 @@
+        setLoading(true);
+
+        try {
+            const result = await flowState.submitCode(code);
```
Does it work? the flowState should be a state, not state handler.

shenj replied:
if (flowState instanceof SignInCodeRequired) {
    **_const handler = AuthFlowStateHandlerFactory.create(flowState);_**
    const result = await handler.submitCode(code);
    ...
}

albanx replied:
will test this evening with the above suggestion

albanx replied:
The above suggested code is for SignIn. This is SignUp code which is not completed yet.
Why I have to build the handler `const handler = AuthFlowStateHandlerFactory.create(flowState);` ?

shenj replied:
I talked it with you previous. Two reasons:
1. we have the circular dependency if we put the handler inside the auth flow result.
2. I think "sharing state" makes more sense than "sharing state handler" cross the views.

shenj replied:
Of course, we can reconsider this if you have a better solution.

albanx replied:
I am going to understand the issue with the circular dependency and try to solve it.

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-up/page.tsx:
```diff
@@ -1,10 +1,117 @@
+    const [email, setEmail] = useState("");
+    const [code, setCode] = useState("");
+    const [error, setError] = useState("");
+    const [flowState, setFlowState] = useState<any>(null);
```
Instead of using any here, It can set as `SignInCompleted | SignInFailed | SignInCodeRequired | SignInPasswordRequired`.

albanx replied:
This points out the point I was making, in terms of S of SOLID. 
flowState has many responsibilities and if we update the SDK with a new state, the engineer have to update his integration code, resulting in breaking the backward compatibility exactly because it is breaking the O of open close.

albanx replied:
I tried this does not work

shenj replied:
Why it doesn't work? Also we can consider using the AuthFlowStateBase since it is better than any.

shenj replied:
Similar suggestion: can we use the base class `AuthFlowStateBase` or `SignUpCompleted | SignUpFailed | SignUpCodeRequired | SignUpPasswordRequired | SignUpAttributesRequired` to replace `any`?

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-up/page.tsx:
```diff
@@ -1,10 +1,117 @@
+        try {
+            const result = await flowState.submitCode(code);
+
+            if (result.error) {
```
Same here. I think it is better to use Failed state check.

shenj commented on yarn.lock:
As we discussed, please don't check in this yarn file.

albanx replied:
for the time being I can add it to the local sample git ignore, to garantee at least that we do not interfeer with the existing msal stuff

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-up/page.tsx:
```diff
@@ -1,10 +1,117 @@
+        setLoading(true);
+
+        try {
+            const app = await CustomAuthPublicClientApplication.create(customAuthConfig);
```
I will suggest using a tool (e.g., Context API) to share the app and flow state cross the views. I know you don't like using Context API, so you can choose whatever tool you like for implementing this. In the sample, we need to show how to split the views when using the SDK.

albanx replied:
There is no need for using the context. In the samples I am solving the issue just with prop drilling to split the view. One other can be with high order components ect...
There are many ways to solve the view split, the question is, what is the one most flexible and less code that can fit in any shape app ? we had the discussion.

shenj replied:
Without such implementation, how to share the the app and state cross the different view and scenario (share the app cross signin signup and sspr)?

axhaferllari committed:
Address comments

albanx committed:
Merge pull request #7625 from AzureAD/custom-auth/alban/angular-sample

Angular Sample skeleton

axhaferllari committed:
Merge branch 'custom-auth/main' into custom-auth/alban/react-next-sample-app-signin

shenj committed:
Merge remote-tracking branch 'origin/dev' into custom-auth/main

shenj committed:
Rebuild

yongdiw committed:
Add doc for all public components and update index.ts (#7634)

This PR is to
1. Add doc to all public components
2. Update index.ts for missing classes

shenj committed:
Fix the some unit tests which caused the open handles issue (#7619)

axhaferllari committed:
[Fix] Fix development on local and remove not needed checks, improved correlationid agnostic fetch client

axhaferllari committed:
Merge branch 'custom-auth/main' into custom-auth/alban/react-next-sample-app-signin

# Conflicts:
#	package-lock.json

axhaferllari committed:
Fix extension module for discovery

axhaferllari committed:
add Sign Up type cast

axhaferllari committed:
fix linting issue

albanx committed:
Merge pull request #7639 from AzureAD/custom-auth/alban/fix-for-dev

[Fix] Fix development on local and remove not needed checks, improved…

axhaferllari committed:
Merge branch 'custom-auth/main' into custom-auth/alban/react-next-sample-app-signin

axhaferllari committed:
complete signup

axhaferllari committed:
Add Reset password sample React JS

shenj commented:
Do you think we can move this sample app into the repo Kaushik suggested? I am planning the create the PR to MSAL for JS team soon for checking in the changes into the dev branch. If it is possible, it is better to exclude the sample app in that PR.

albanx commented:
> Do you think we can move this sample app into the repo Kaushik suggested? I am planning the create the PR to MSAL for JS team soon for checking in the changes into the dev branch. If it is possible, it is better to exclude the sample app in that PR.

Yes I have moved them partially, not the last version. Otherwise I cannot run them in the repo until the package is published.

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-in/page.tsx:
```diff
@@ -1,10 +1,117 @@
+                    setError("Incorrect password");
+                } else {
+                    setError("An error occurred during sign in");
+                }
```
Maybe I misunderstood, but any reason why you didn't use the `handleError` function you defined in the utils/index.ts?

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-in/page.tsx:
```diff
@@ -1,10 +1,117 @@
+        setLoading(true);
+
+        try {
+            const result = await flowState.submitCode(code);
```
As discussed earlier, can you double check whether `flowState.submitCode(code)` is working?

albanx replied:
I have ran the sign in different times, it works, unless there has been some recent updates.

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-in/page.tsx:
```diff
@@ -1,10 +1,117 @@
+    const [password, setPassword] = useState("");
+    const [code, setCode] = useState("");
+    const [error, setError] = useState("");
+    const [flowState, setFlowState] = useState<any>(null);
```
Do you think it is better to use the state base class `AuthFlowStateBase` to replace `any`?

shenj replied:
Or use `SignInCompleted | SignInFailed | SignInCodeRequired | SignInPasswordRequired`?

albanx replied:
Can be `AuthFlowStateBase `, the later one I really do not recommend. This is how developer experience will look with the latest:
`const [flowState, SetFlowState] = useState<` checking docs

I see `SignInCompleted `, perfect
`const [flowState, SetFlowState] = useState<SignInCompleted>(null)`

Ahh wait there is more
`const [flowState, SetFlowState] = useState<SignInCompleted | SignInFailed>(null)`

no wait more types in or

`const [flowState, SetFlowState] = useState<SignInCompleted | SignInFailed | SignInCodeRequired >(null)`
(how many possible types in OR I should have, I have to check all of them and find out by reading a long boring documentation)


1 month after the SDK releseas version 2.0 ... Ah crap I have to update all my code again

`const [flowState, SetFlowState] = useState<SignInCompleted | SignInFailed | SignInCodeRequired | **ANewSignInState** >(null)`


```
if (flowState instanceOf SignInCompleted)
if (flowState instanceOf SignInFailed )
if (flowState instanceOf SignInCodeRequired )
**if (flowState instanceOf ANewSignInState)**
```

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-up/page.tsx:
```diff
@@ -1,10 +1,117 @@
+        setLoading(true);
+
+        try {
+            const app = await CustomAuthPublicClientApplication.create(customAuthConfig);
```
IMHO, the app can be shared cross the whole application for sign-in, sign-up and sspr. Is it possible we can use the single instance of app in this sample app?

albanx replied:
There is no point and need to share the same app instance for different flows that are not connected in the lifecycle of an app. Imagine Sign Up , Sign In and SSPR as different flows that are separated in time and usage , as it happens in every app (a user does not go through the all flows in the same app cycle). Making this flows sharing the same instance it kind of tight them together for no reason.

shenj commented on samples/msal-custom-auth-samples/react-sample-nextjs/src/app/sign-up/page.tsx:
```diff
@@ -1,10 +1,117 @@
+        <div style={styles.container}>
+            <h2>Sign Up</h2>
+            {flowState?.type === SignUpState.CodeRequired ? (
+                <CodeForm onSubmit={handleCodeSubmit} code={code} setCode={setCode} loading={loading} />
```
May I ask why we don't have the PasswordForm?

albanx replied:
This is SignUp with OTP does not asks for password which I based on the past samples. That is the configuration of the sample tenant. I can add the password field but it will be optional

axhaferllari committed:
Fix Sign In with OTP

axhaferllari committed:
added Sign up with password

axhaferllari committed:
remove tenant URLs

axhaferllari committed:
Add comment
